### ISSUE TITLE
Memory leak

### ISSUE THREAD

**Issue by armed10** [stellar/stellar-core#1033]
Continuesly sending transactions to a stellar-core, either on a standalone config or in a network, results in slow increase of memory usage (RAM). This increase is about 1 Gigabyte of ram per 1 million operations sent to the machine. On our modest machines, the laptop eventually crashes due to the RAM and swap being full.

This issue was discovered in a 0.4.1 release but is still an issue in a pull from the current master branch.

**Comment by vogel** [comment]
Issue confirmed. I'll look into that.

**Comment by vogel** [comment]
My results after testing of "generateload" feature on local cluster on 4 stellar-core instances with different cache settings:

- generator instance was using up to 1.6 GB of memory (with cache full)
- normal instance was using up to 1.2 GB of memory (with cache full)
- instance with cache size of transaction reduced was using about 670 MB of memory (with cache full)
- instance with all cache size reduced was using about 620 MB of memory (with cache full)

It does not seem there is memory leak connected with generateload feature.
I'll try now with stellar-core -c tx?blob=...

May I ask you how much memory was stellar-core using on your laptop and how did you sent transaction to it?

**Comment by vogel** [comment]
Forgot to mention: above memory values are measured under valgrind/massif tool, so are much larger than for normal stellar-core usage.

Also previous confirmation of issue is invalid - the memory usage patterns I noticed were connected with filling up internal cache structures.

**Comment by vogel** [comment]
Not found in case of executing transactions with stellar-core -c tx?blob=...
I will enable writing history in next try - maybe it is taking all the memory.

**Comment by vogel** [comment]
There is no problem with memory leak, but with some assumptions that we made.

For example, by default, we keep 10 000 old TxSets in cache, each can have 50 transactions (this value can be changed in configuration file). Transactions are typically pretty small, about 1-2 kb of data. This allows for as much as 1GB data in memory.

However, the bigged transaction that I was able to make was about 42 kb of size. It makes upper limit of TxSets cache in realm of 20GB of data. If we extend limit of transactions per ledger we can extend this memory requirements as well.

I'll talk with @jedmccaleb how to prevent that. One solution is to remove too old transactions sets from the cache (I've done poc on that, working pretty well), but with one problem: I don't know if txsets can be reused between ledgers (txsets do not store hash of ledger inside).

We also could make TxSets cache size depending of max transaction per ledger. If we increase the latter from 50 to 500, we should make TxSets smaller (or get more RAM).
